### PR TITLE
allow react memo component can set displayName multiple times

### PR DESCRIPTION
**Pull request by iChenLei**
## TL;DR
1. React memo can set displayName ( begin at React v17 https://github.com/facebook/react/issues/18026#issuecomment-675900452
2. React memo displayName's weight is lower than inner wrapped component's displayName https://github.com/facebook/react/pull/15313#discussion_r429769752
3. Change the current behavior to hoist memo displayName's show weight and allow set memo displayName multiple times.

## React displayName
question: can we set component displayName multiple times?
> codesandbox link 👉    https://codesandbox.io/s/react-17-memo-displayname-forked-yk4u0
<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>
 
answer: ⬇️
1. React.forwardRef 👌
2. Normal Function Component 👌
3. Class Component 👌
4. React.memo 🙅‍♂️

#### React.forwardRef displayName behavior
```js
 //  packages/react/src/ReactForwardRef.js
  const elementType = {
    $$typeof: REACT_FORWARD_REF_TYPE,
    render,    👈🏻   // render: render not type: type
  };
  if (__DEV__) {
    let ownName;
    Object.defineProperty(elementType, 'displayName', {
      enumerable: false,
      configurable: true,
      get: function() {
        return ownName;
      },
      set: function(name) {
        ownName = name;
        if (render.displayName == null) {
          render.displayName = name;  👈🏻   //  only effect render property
        }
      },
    });
  }
  return elementType;
}
```
#### React forwardRef Component actually

<div style="display:flex;justify-content:center;width=650px;">
		<p>
	    </p>
</div>

You can modilfy `displayName` multiple times, because it's standalone getter/setter property.

#### React devtools getDisplayName logic
```js
export function getDisplayName(
  type: Function,
  fallbackName: string = 'Anonymous',
): string {
  const nameFromCache = cachedDisplayNames.get(type);
  if (nameFromCache != null) {
    return nameFromCache;
  }
  let displayName = fallbackName;
  // The displayName property is not guaranteed to be a string.
  // It's only safe to use for our purposes if it's a string.
  // github.com/facebook/react-devtools/issues/803
  if (typeof type.displayName === 'string') {    👈🏻   // type.displayName
    displayName = type.displayName;
  } else if (typeof type.name === 'string' && type.name !== '') {
    displayName = type.name;
  }
  cachedDisplayNames.set(type, displayName);
  return displayName;
}
```

## PR for what ?
1. For react ecosystem third party library.  https://github.com/mobxjs/mobx/issues/2721#issue-788324435
2. Make `React.forwardRef` and `React.memo` have same behavior
> @eps1lon  's previous comment for this: 
This test was added in PR 18925 but I don't think this should've been the expected behavior (at bvaughn.
For React.memo we ignore displayName on the wrapper if we have a displayName on the wrapped component.
For React.forwardRef we honor displayName on the wrapper even if we have displayName on the wrapped component.
I think it is confusing that React.memo and React.forwardRef behave differently with regard to displayName.

## Test Plan
Add a new unit test and modify a exist unit test. Need @bvaughn and @eps1lon check it.

### append comment
welcome any advices helpful, I'm not ReactJS expert so maybe I do something wrong.

### CODE REVIEW

**Review by rheh — APPROVED** [review]

**Comment by iChenLei** [comment]
@bvaughn @eps1lon Can someone review this pull request? If it's not ok I will close it , Thanks.

**Comment by bvaughn** [comment]
I'll add myself as a reviewer to reminder myself at some point in the future but I'm pretty swamped right now so I'm not committing to look any time soon.

It doesn't hurt anything to leave the PR open so I suggest you just do that. :)

**Comment by iChenLei** [comment]
Sorry, it's my fault. Thanks for your advices.

**Comment by codeBelt** [comment]
Is there a timeline when this might be merged?

**Review by ccheney — APPROVED** [review]
LGTM

File: packages/react/src/ReactMemo.js [hunk 1]
```diff
@@ -36,9 +36,14 @@ export function memo<Props>(
...
-        ownName = name;
-        if (type.displayName == null) {
+        if (typeof name === 'string') {
+          ownName = name;
           type.displayName = name;
+        } else {
+          console.error(
+            "%s: is not valid displayName type, React memo's displayName should be a string",
+            typeof name,
+          );
```
Comment (bvaughn) [inline packages/react/src/ReactMemo.js:46]: Why not follow the exact behavior of `forwardRef` here?
https://github.com/iChenLei/react/blob/d38a651fbea3bae55b572afbae1f96e836dd27f6/packages/react/src/ReactForwardRef.js#L51-L64

Specifically, only set `type.displayName = name` if it's not already `type.displayName === null`.

Comment (iChenLei) [reply packages/react/src/ReactMemo.js:46]: https://github.com/facebook/react/blob/a0d6b155dc3a182162b9f91baac33df39d7919df/packages/react-devtools-shared/src/backend/renderer.js#L379-L392

If we set ownName we get the ownName(`ownName`) as displayName and otherwise we use type.render's displayName or name.

https://github.com/facebook/react/blob/a0d6b155dc3a182162b9f91baac33df39d7919df/packages/react-devtools-shared/src/backend/renderer.js#L414-L419

Yep, you are right, we need change the forwardRef logic as same as memoComponent.

Comment (iChenLei) [reply packages/react/src/ReactMemo.js:46]: We just want allow user set displayName multi time, the common scenario is that user use third party react library such as mobx-react, mobx-react use hoc to wrap component and use `memo` to optimize rende performance. and now user can't set custom displayName to debug . for example (https://github.com/mobxjs/mobx/issues/2721

**Comment by iChenLei** [comment]
memo-displayName

For example, you can't reset displayName for debug. But if we change the original behavior we will break `should honor a inner displayName if set on the wrapped function` from `packages/react-reconciler/src/__tests__/ReactMemo-test.js`. I am really confused now.  so sad.

File: packages/react/src/ReactMemo.js (hunk 1 above)
Comment (bvaughn) [reply packages/react/src/ReactMemo.js:46]: My point was that I don't understand why we were making the implementation of `forwardRef` and `memo` different.

TBH I'm not sure that _either one_ modifying `type.displayName` is a good idea b'c that's a side effect. We do it because it helps with component stacks but I'm not sure that's the best workaround there either.

For instance:

```js
function MyComponent() {
  // ...
}

const MyMemoizedComponent = React.memo(MyComponent);
MyMemoizedComponent.displayName = 'Blah';

// Now if I render <MyComponent> it will also have a displayName of "blah"
// That doesn't seem good.
```

It's true that `forwardRef` and `memo` currently have different behaviors in component stacks and that's not good either. To me, this indicates that the right place to make this change is where we read the names rather than here. So I'm talking about `getComponentNameFromType` 😄 

Maybe it would be clearer for me to post a PR with an alternate idea for this.

Comment (bvaughn) [reply packages/react/src/ReactMemo.js:46]: See #21392

**Comment by iChenLei** [comment]
closed, see bvaughn's alternative PR  #21392
